sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -600,20 +600,27 @@ public Duration getAllowedTimestampSkew() {
   @Target(ElementType.METHOD)
   public @interface FinishBundle {}
 
-
   /**
-   * Annotation for the method to use to clean up this instance after processing bundles of
-   * elements. No other method will be called after a call to the annotated method is made.
-   * The method annotated with this must satisfy the following constraint:
+   * Annotation for the method to use to clean up this instance before it is discarded. No other
+   * method will be called after a call to the annotated method is made.
+   *
+   * <p>Note that calls to the annotated method are best effort, and may not occur for arbitrary
```

Comment:
What would it cost to not make it best effort?
Basically an API which is "best effort" is quite pointless and since beam must not assume the JVM is not reused (it is by some runners) then beam should ensure the transforms can release whatever they used so enforce the call to teardown whatever happent, no?

Reply:
It is **impossible** to make it guaranteed, as a machine may fail at any time for any reason. Users should _never_ depend on invocation of teardown to produce results.

Teardown being reliable when there are no failures is possible, but that's a combination of per-runner and per-harness behavior, which I've filed https://issues.apache.org/jira/browse/BEAM-3704 to enable

Reply:
Ok so the javadoc should mention it can be missed due to a crash but except that it is guaranteed. Any code can be dropped due to a crash like the finish bundle you recommand in this javadoc so it is quite misleading now and makes the api not that helpful from a user perspective :(.

Reply:
I assume the purpose here is to guide the end user to use `tearDown` method appropriately.  I think it is better to assume the user is beginner and does not know much about how the the job is executed. May be better to emphasis 'best effort' more and (say something to the effect of 'usually invoked during a orderly shutdown'), possibly remove the phrase 'may not occur for arbitrary reasons'.  Just thinking aloud.. non-binding review comment.